# 문서 군집화(Document Clustering)

* 비슷한 텍스트 구성의 문서를 군집화 - 동일한 군집에 속하는 문서를 같은 카테고리 소속으로 분류
* 비지도 학습 기반

### 데이터 세트를 이용한 문서 군집화 수행
* 데이터 세트 : https://archive.ics.uci.edu/ml/datasets/Opinosis+Opinion+%26frasl%3B+Review


In [1]:
import pandas as pd
import glob, os

path = r'/content/drive/MyDrive/파이썬_머신러닝_완벽가이드/08_Text_Analytics/Text'
all_files = glob.glob(os.path.join(path, "*.data"))
all_files.sort()
filename_list = []
opinion_text = []

for file_ in all_files:
    df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
    filename_ = file_.split('/')[-1]
    filename = filename_.split('.')[0]
    filename_list.append(filename)
    opinion_text.append(df.to_string())

document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

filename                                       opinion_text
0   accuracy_garmin_nuvi_255W_gps                                                ...
1  bathroom_bestwestern_hotel_sfo                                                ...
2      battery-life_amazon_kindle                                                ...
3      battery-life_ipod_nano_8gb                                                ...
4     battery-life_netbook_1005ha                                                ...

각 파일 이름(filename) 자체만으로 의견(opinion)의 텍스트(text)가 어떠한 제품/서비스에 대한 리뷰인지 잘 알 수 있다.

문서를 TF-IDF 형태로 피처 벡터화한다.

In [2]:
from nltk.stem import WordNetLemmatizer
import nltk
import string

nltk.download('all')

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package biocreative_ppi is already up-to-date!
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Package brown is already up-to-date!
[nltk_data]    | Downloading package brown_tei to /root/nltk_data...
[nltk_data]    |   Package brown_tei is already up-to-date!
[nltk_data]    | Downloading package cess_cat to /root/nltk_data...
[nltk_data]    |   Package cess_cat is already up-to-date!
[nltk_data]    | Downloading package cess_esp to /root/nltk_data...
[nltk_data]    |   Package cess_esp is already up-to-date!
[nltk_data]    | Downloading packag

In [3]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english', ngram_range=(1,2), min_df=0.05, max_df=0.85 )

#opinion_text 컬럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

/usr/local/lib/python3.7/dist-packages/sklearn/feature_extraction/text.py:401: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['ha', 'le', 'u', 'wa'] not in stop_words.
  % sorted(inconsistent)


### K-평균 군집화 적용

In [4]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행.
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

각 데이터별로 할당된 군집의 레이블을 파일명과 파일 내용을 가지고 있는 document_df DataFrame에 'cluster_label' 칼럼을 추가하여 저장.

각 파일명은 의견 리뷰에 대한 주제를 나타낸다.

In [5]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  ... cluster_label
0   accuracy_garmin_nuvi_255W_gps  ...             2
1  bathroom_bestwestern_hotel_sfo  ...             0
2      battery-life_amazon_kindle  ...             1
3      battery-life_ipod_nano_8gb  ...             1
4     battery-life_netbook_1005ha  ...             1

[5 rows x 3 columns]

In [6]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  ... cluster_label
1   bathroom_bestwestern_hotel_sfo  ...             0
30         room_holiday_inn_london  ...             0
31     rooms_bestwestern_hotel_sfo  ...             0
32         rooms_swissotel_chicago  ...             0

[4 rows x 3 columns]

Cluster #0 - 호텔 리뷰

In [7]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  ... cluster_label
2    battery-life_amazon_kindle  ...             1
3    battery-life_ipod_nano_8gb  ...             1
4   battery-life_netbook_1005ha  ...             1
19      keyboard_netbook_1005ha  ...             1
26   performance_netbook_1005ha  ...             1
41     size_asus_netbook_1005ha  ...             1
42          sound_ipod_nano_8gb  ...             1
44               speed_windows7  ...             1

[8 rows x 3 columns]

Cluster #1 - 킨들, 아이팟, 넷북 등의 포터블 전자기기에 대한 리뷰

In [8]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  ... cluster_label
0     accuracy_garmin_nuvi_255W_gps  ...             2
5             buttons_amazon_kindle  ...             2
8   directions_garmin_nuvi_255W_gps  ...             2
9      display_garmin_nuvi_255W_gps  ...             2
10    eyesight-issues_amazon_kindle  ...             2
11                features_windows7  ...             2
12              fonts_amazon_kindle  ...             2
23         navigation_amazon_kindle  ...             2
33   satellite_garmin_nuvi_255W_gps  ...             2
34      screen_garmin_nuvi_255W_gps  ...             2
35             screen_ipod_nano_8gb  ...             2
36            screen_netbook_1005ha  ...             2
43       speed_garmin_nuvi_255W_gps  ...             2
48     updates_garmin_nuvi_255W_gps  ...             2
49              video_ipod_nano_8gb  ...             2
50       voice_garmin_nuvi_255W_gps  ...             2

[16 rows x 3 columns]

Cluster #2 - Cluster #1과 비슷하게 킨들, 아이팟, 넷붓이 군집에 포함되어 있지만, 주로 차량용 네비게이션에 대한 리뷰(xxx_garmin_nuvi_xxx)

In [9]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  ... cluster_label
13          food_holiday_inn_london  ...             3
14           food_swissotel_chicago  ...             3
15       free_bestwestern_hotel_sfo  ...             3
20   location_bestwestern_hotel_sfo  ...             3
21      location_holiday_inn_london  ...             3
24    parking_bestwestern_hotel_sfo  ...             3
27              price_amazon_kindle  ...             3
28         price_holiday_inn_london  ...             3
38    service_bestwestern_hotel_sfo  ...             3
39       service_holiday_inn_london  ...             3
40  service_swissotel_hotel_chicago  ...             3
45      staff_bestwestern_hotel_sfo  ...             3
46          staff_swissotel_chicago  ...             3

[13 rows x 3 columns]

Cluster #3 - 킨들(kindle) 리뷰가 한 개씩 썩여있어 아쉽지만, Cluster #0과 같이 대부분 호텔에 대한 리뷰

In [10]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  ... cluster_label
6        comfort_honda_accord_2008  ...             4
7        comfort_toyota_camry_2007  ...             4
16   gas_mileage_toyota_camry_2007  ...             4
17      interior_honda_accord_2008  ...             4
18      interior_toyota_camry_2007  ...             4
22       mileage_honda_accord_2008  ...             4
25   performance_honda_accord_2008  ...             4
29       quality_toyota_camry_2007  ...             4
37         seats_honda_accord_2008  ...             4
47  transmission_toyota_camry_2007  ...             4

[10 rows x 3 columns]

Cluster #4 - 토요타(Toyota)와 혼다(Honda) 등의 자동차에 대한 리뷰

전반적으로 군집화된 결과를 보면 군집 개수가 약간 많게 설정되어 세분된 것 같다. 군집 중심 개수를 5개에서 3개로 줄여 군집화를 수행.

In [11]:
# 3개의 집합으로 군집화 
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  ... cluster_label
0     accuracy_garmin_nuvi_255W_gps  ...             0
48     updates_garmin_nuvi_255W_gps  ...             0
44                   speed_windows7  ...             0
43       speed_garmin_nuvi_255W_gps  ...             0
42              sound_ipod_nano_8gb  ...             0
41         size_asus_netbook_1005ha  ...             0
36            screen_netbook_1005ha  ...             0
35             screen_ipod_nano_8gb  ...             0
34      screen_garmin_nuvi_255W_gps  ...             0
33   satellite_garmin_nuvi_255W_gps  ...             0
27              price_amazon_kindle  ...             0
26       performance_netbook_1005ha  ...             0
49              video_ipod_nano_8gb  ...             0
23         navigation_amazon_kindle  ...             0
19          keyboard_netbook_1005ha  ...             0
50       voice_garmin_nuvi_255W_gps  ...             0
9      display_garmin_nuvi_255W_gps  ...             0
2        battery-life_amazon_kindle  ...             0
3        battery-life_ipod_nano_8gb  ...             0
4       battery-life_netbook_1005ha  ...             0
5             buttons_amazon_kindle  ...             0
12              fonts_amazon_kindle  ...             0
11                features_windows7  ...             0
10    eyesight-issues_amazon_kindle  ...             0
8   directions_garmin_nuvi_255W_gps  ...             0
47   transmission_toyota_camry_2007  ...             1
37          seats_honda_accord_2008  ...             1
6         comfort_honda_accord_2008  ...             1
7         comfort_toyota_camry_2007  ...             1
16    gas_mileage_toyota_camry_2007  ...             1
25    performance_honda_accord_2008  ...             1
17       interior_honda_accord_2008  ...             1
18       interior_toyota_camry_2007  ...             1
22        mileage_honda_accord_2008  ...             1
29        quality_toyota_camry_2007  ...             1
1    bathroom_bestwestern_hotel_sfo  ...             2
46          staff_swissotel_chicago  ...             2
45      staff_bestwestern_hotel_sfo  ...             2
14           food_swissotel_chicago  ...             2
20   location_bestwestern_hotel_sfo  ...             2
21      location_holiday_inn_london  ...             2
30          room_holiday_inn_london  ...             2
38    service_bestwestern_hotel_sfo  ...             2
13          food_holiday_inn_london  ...             2
24    parking_bestwestern_hotel_sfo  ...             2
28         price_holiday_inn_london  ...             2
15       free_bestwestern_hotel_sfo  ...             2
32          rooms_swissotel_chicago  ...             2
31      rooms_bestwestern_hotel_sfo  ...             2
39       service_holiday_inn_london  ...             2
40  service_swissotel_hotel_chicago  ...             2

[51 rows x 3 columns]

* Cluster #0 - 포터블 전자기기 리뷰
* Cluster #1 - 자동차 리뷰
* Cluster #2 - 호텔 리뷰

### 군집별 핵심 단어 추출
* 각 군집에 속한 문서는 핵심 단어를 주축으로 군집화되어 있을 것이다. 이 핵심 단어를 추출해본다.

* KMeans 객체는 각 군집을 구성하는 단어 피처가 군집의 중심(Centroid)을 기준으로 얼마나 가깝게 위치해 있는지 cluster_centers_라는 속성으로 제공. 
* cluster_centers_는 배열 값으로 제공되며, 행은 개별 군집을, 열은 개별 피처를 의미.
* 각 배열 내의 값은 개별 군집 내의 상대 위치를 숫자 값으로 표현한 일종의 좌표 값이다.

In [12]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape :',cluster_centers.shape)
print(cluster_centers)

cluster_centers shape : (3, 4611)
[[0.01005322 0.         0.         ... 0.00706287 0.         0.        ]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]
 [0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]]


cluster_centers_는 (3, 4611) 배열이다. 이는 군집이 3개, word 피처가 4611개로 구성.

각 행의 배열 값은 각 군집 내의 4611개 피처의 위치가 개별 중심과 얼마나 가까운가를 상대값으로 나타낸 것이다. 0에서 1까지의 값을 가질 수 있으며 1에 가까울수록 중심과 가까운 값을 의미한다.

In [13]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상대값, 대상 파읿명을 반환하는 함수
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}

    # cluster_centers_array의 값이 큰 순으로 정렬된 인덱스 값을 반환
    # ndarray의 argsort()[:, ::-1]을 이용하여 cluster_centers 배열 내 값이 큰 순으로 정렬된 위치 인덱스 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]

    # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상대값, 대상 파일멸 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num

        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]

        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상대값 구함
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()

        # cluster_details 딕셔너리 객체에 개별 군집별 핵심단어와 중심위치 상대값, 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames
        
    return cluster_details  

get_cluster_details()를 호출하면 개별 군집번호, 핵심 단어, 핵심단어 중심 위치 상대값, 파일명 속성 값 정보를 담은 dictionary를 반환한다.

In [14]:
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

get_cluster_detail() 호출 시 인자는 KMeans 군집화 객체, 파일명 추출을 위한 document_df인 DataFrame, 핵심 단어 추출을 위한 피처명 리스트, 전체 군집 개수 그리고 핵심 단어 추출 개수이다. 피처명 리스트는 TF-IDF 변환된 tfidf_vect 객체에서 get_feature_names()로 추출.

In [15]:
feature_names = tfidf_vect.get_feature_names()
cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,\
                                  feature_names=feature_names, clusters_num=3, top_n_features=10 )
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['accuracy_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'battery-life_ipod_nano_8gb', 'battery-life_netbook_1005ha', 'buttons_amazon_kindle', 'directions_garmin_nuvi_255W_gps', 'display_garmin_nuvi_255W_gps']
####### Cluster 1
Top features: ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 : ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'gas_mileage_toyota_camry_2007', 'interior_honda_accord_2008', 'interior_toyota_camry_2007', 'mileage_honda_accord_2008', 'performance_honda_accord_2008']
####### Cluster 2
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'food_swissotel_chicago', 'free_bestwestern_h

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


* Cluster #0 - 포터블 전자제품 리뷰 : 'screen', 'battery', 'life' 등과 같은 화면과 배터리 수명 등이 핵심 단어로 군집화.
    * 모바일형, 엔터테인먼트용 전자제품의 경우 화면 크기와 배터리 수명이 주요 관심사로 보인다.
* Cluster #1 - 자동차 리뷰 : 'interior', 'seat', 'mileage', 'comfortable' 등과 같은 실내 인테리어, 좌석, 연료 효율 등이 핵심 단어로 군집화. 
    * 토요타, 혼다와 같은 일본 자동차의 경우 실내 인테리어와 연료 효율, 편안함이 주요 관심사로 보인다.
* Cluster #2 - 호텔 리뷰 : 'room', 'hotel', 'service', 'location' 등과 같은 방과 서비스 등이 핵심 단어로 군집화.
    * 호텔 리뷰의 경우 방의 크기나 청소 상태, 직원들의 서비스, 위치 등이 주요 관심사로 보인다.